py-polars/tests/unit/functions/as_datatype/test_duration.py

```diff
@@ -143,6 +143,68 @@ def test_add_duration_3786() -> None:
     }
 
 
+@pytest.mark.parametrize(
+    ("dt_time_unit", "duration_time_unit", "expected_time_unit"),
+    [
+        ("ms", "ms", "ms"),
+        ("ms", "us", "us"),
+        ("us", "ms", "us"),
+        ("us", "us", "us"),
+        ("ms", "ns", "ns"),
+        ("us", "ns", "ns"),
+        ("ns", "ms", "ns"),
+        ("ns", "us", "ns"),
+        ("ns", "ns", "ns"),
+    ],
+)
+def test_datetime_add_duration_subsecond(
+    dt_time_unit: TimeUnit, duration_time_unit: TimeUnit, expected_time_unit: TimeUnit
+) -> None:
+    dt_value = {
+        "ms": 123,
+        "us": 123_456,
+        "ns": 123_456_789,
+    }[dt_time_unit]
+
+    df_datetimes = pl.DataFrame({"ts": [f"2020-01-01T00:00:00.{dt_value}"]}).select(
+        pl.col("ts").str.to_datetime(time_unit=dt_time_unit)
+    )
+
+    verbose_duration_unit = {
+        "ms": "milliseconds",
+        "us": "microseconds",
+        "ns": "nanoseconds",
+    }[duration_time_unit]
+
+    duration = pl.duration(
+        **{verbose_duration_unit: 1},
+        time_unit=duration_time_unit,
+    )
+
+    df_add_right = df_datetimes.select((pl.col("ts") + duration).alias("ts"))
+    df_add_left = df_datetimes.select((duration + pl.col("ts")).alias("ts"))
+
+    multipliers = {
+        "ms": 1_000_000,
+        "us": 1_000,
+        "ns": 1,
+    }
+    dt_multiplier = multipliers[dt_time_unit]
+    offset = multipliers[duration_time_unit]
```

Comment:
this is looking quite complicated for a test, is there a simpler way to write it?